pubtools/exodus/task.py

```diff
@@ -88,6 +91,15 @@ def main(self):
         with task_context():
             # setup the logging as required
             self._setup_logging()
-
-            self.run()
+            if self.exodus_enabled():
+                self.run()
+            else:
+                LOG.warning(
```

Comment:
I don't think this (warn and then exit successfully) is the right behavior.  In JIRA I wrote the expected behavior as "The command fails".

I believe it's right for this to be inconsistent with other bits of code like the hooks, because the scenario is quite different.  Other cases like the hooks, or the pub provision command with multiple backends, uses EXODUS_ENABLED to choose between different code paths.  The task in progress can still do something meaningful if exodus support is disabled.

pubtools-exodus-push is different though.  That's a command which is explicitly for pushing to exodus.  If you call it while exodus support is disabled or not configured then some mistake has been made; we should fail in order to highlight the mistake, rather than succeed and output a warning which might be ignored.

The reason I filed this issue in the first place is that someone had configured their Pub environment with push_staged_cdn_backend=exodus but had not set any of the `EXODUS_*` config.  It was wrong for the Pub admin to do that, but it was also wrong that pubtools-exodus-push gave a confusing crash.  If pubtools-exodus-push instead succeeded, but didn't actually do any push, that would still be wrong as the Pub admin might not realize the mistake.  Failing is the user-friendly thing to do here.

Reply:
Sorry, quick clarification please:

I interpreted your comment in #11 ...
> I think the doc & **behavior for [EXODUS_ENABLED]** should be something like: pubtools-exodus can optionally extend the behavior of other projects in the pubtools family, causing them to integrate with exodus-gw. This is enabled if and only if EXODUS_ENABLED is set.
> 
> And then it might make sense for the header to be something like **"Enabling pubtools-exodus hooks"** or similar. 

_[emphasis my own]_

... to mean that `EXODUS_ENABLED` should be reserved for enabling the pubtools-exodus pulp hooks, and shouldn't apply to the `pubtools-exodus-*` commands. Is that correct? Should we always assume that "exodus is enabled" when invoking the `pubtools-exodus-push` command, given that we already toggle its use via the `push_staged_cdn_backend=exodus` target setting?

If the answer is "no", please do not review this commit yet :).

Reply:
I've been operating under the assumption that *all* actions targeting Exodus should first consult the Pub target's settings. I'm unfamiliar with the `push_staged_cdn_backend` setting but the intent of the `EXODUS_ENABLED` setting var was to have a single on/off switch. It'd be really confusing if that's being overridden in various places.

Reply:
`push_staged_cdn_backend` was born in https://issues.redhat.com/browse/RHELDST-10120:

[regarding the use of separate RhsmPushStaged and ExodusPushStaged classes]
> That means, unlike in the PulpPushStaged case, we won't have any support for publishing both to NetStorage and exodus at the same time in a single task. If you set up a kickstart or ostree push with Pub, you are going to push only to NetStorage or only to Exodus. In the worst case scenario where you really need both, then either you need to do one push, then have someone adjust the target settings, then do another push; or we need to set up multiple targets to be used with that push-staged use-case.

>>  Altering the push target to EXODUS_ENABLED=false sounds dangerous (without first instating some kind of release freeze). Otherwise, other content pushed during that time wouldn't land in the expected destinations.

> Yes. This could be mitigated by not using EXODUS_ENABLED for this after all but instead using a target
setting dedicated to this specific case (e.g. "push_staged_cdn_backend": "cdn" or "exodus"), so flipping
the setting wouldn't affect tasks using Pulp.

I guess re-reading this conversation answers my question re: whether to use EXODUS_ENABLED here, but I also don't want to cause confusion, so I'll wait for confirmation.